closed/src/java.base/share/native/libjncrypto/NativeCrypto.c

```diff
@@ -380,30 +401,91 @@ JNIEXPORT jboolean JNICALL Java_jdk_crypto_jniprovider_NativeCrypto_isOpenSSLFIP
-    crypto_library = load_crypto_library(trace);
-    if (NULL == crypto_library) {
-        if (trace) {
-            fprintf(stderr, "Error loading OpenSSL: FAILED TO LOAD OPENSSL CRYPTO LIBRARY\n");
-            fflush(stderr);
+void log_crypto_library_path(jboolean traceEnabled, void *crypto_library, char *message) {
+    if (traceEnabled && (NULL != crypto_library)) {
+
+    #if defined(_AIX)
+        // Initialize the buffer with maximum size for L_GETINFO.
+        char *buffer = (char *) malloc(DLFCN_LDINFO_SIZE);
+        int rc = -1;
+        // Get the list of all object files loaded by this process
+        rc = loadquery(L_GETINFO, buffer, DLFCN_LDINFO_SIZE);
+        if (NULL == buffer) {
+            return;
         }
-        return -1;
+
+        // Parse the list of all object files and print the OPENSSL library path.
+        if (rc == 0) {
+            char *buf = buffer;
+            struct ld_info* cur_info = NULL;
+            do {
+                cur_info = (struct ld_info*) buf;
+                char *path = cur_info->ldinfo_filename;
+                char *member_name = path + strlen(cur_info->ldinfo_filename) + 1;
+                if ((*member_name != '\0') && (strstr(path, "libcrypto") != NULL)) {
+                    fprintf(stdout, "%s: %s(%s)\n", message, path, member_name);
+                    fflush(stdout);
+                    free(buffer);
+                    return;
+                }
+                buf += cur_info->ldinfo_next;
+            } while (cur_info->ldinfo_next != 0);
+        }
+        free(buffer);
+
+    #elif defined(__APPLE__)
+        // Since we know the image we want will always be near the end of the list, start there and go backwards.
+        for (uint32_t i = (_dyld_image_count() - 1); i >= 0; i--) {
+            const char *image_name = _dyld_get_image_name(i);
+
+            // Why dlopen doesn't effect _dyld stuff: if an image is already loaded, it returns the existing handle.
+            void *probe_handle = dlopen(image_name, RTLD_LAZY);
+            dlclose(probe_handle);
+
+            if (crypto_library == probe_handle) {
+                fprintf(stdout, "OpenSSL was loaded from - %s\n", image_name);
+                fflush(stdout);
+                return;
+            }
+        }
+    #elif defined (_WIN32)
+        char path [MAX_PATH];
+        GetModuleFileName(crypto_library, path, MAX_PATH);
+        fprintf(stdout, "OpenSSL was loaded from - %s\n", path);
+        fflush(stdout);
+    #else
+        struct link_map *map = NULL;
+        dlinfo(crypto_library, RTLD_DI_LINKMAP, &map);
+        fprintf(stdout, "OpenSSL was loaded from - %s\n", map->l_name);
+        fflush(stdout);
+    #endif
     }
+}
+
+/* Get the version for the crypto library */
+jlong get_crypto_library_version(jboolean traceEnabled, void *crypto_library, char *message) {
+    typedef const char* OSSL_version_t(int);
+    OSSL_version_t* OSSL_version;
+    const char * openssl_version;
```

Comment:
Finish each sentence with a period.
Return type, opening brace on separate lines.
The function can be `static`.
The final parameter type can be `const char *`.
Initialize each local.
Space before, not after `*`.
```
/* Get the version for the crypto library. */
static jlong
get_crypto_library_version(jboolean traceEnabled, void *crypto_library, const char *message)
{
    typedef const char *OSSL_version_t(int);
    OSSL_version_t *OSSL_version = NULL;
    const char *openssl_version = NULL;
```

Reply:
Fixed.